compiler/src/iree/compiler/Dialect/Util/Transforms/SimplifyGlobalAccesses.cpp

```diff
@@ -78,240 +89,688 @@ static void hoistImmutableLoads(Region &region,
+      for (auto &block : region) {
+        for (auto &containedOp : block) {
+          auto accessor = solver.getElementFor<GlobalAccessor>(
+              *this, Position::forOperation(&containedOp),
+              DFX::Resolution::REQUIRED);
+          bestState ^= accessor;
+        }
+      }
+    }
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "[simplify-globals] region op: ";
+      op->print(llvm::dbgs(), solver.getAsmState());
+      llvm::dbgs() << "; " << bestState;
+      llvm::dbgs() << "\n";
+    });
+    return DFX::inheritStateAndIndicateChange(getState(), bestState);
+  }
+
+  friend class DFX::Solver;
+};
+const char GlobalAccessor::ID = 0;
+
+class GlobalAccessorAnalysis {
+public:
+  explicit GlobalAccessorAnalysis(Operation *rootOp)
+      : explorer(rootOp, TraversalAction::RECURSE,
+                 OpPrintingFlags().elideLargeElementsAttrs().skipRegions()),
+        solver(explorer, allocator) {
+    // TODO: Use op traits or something to ignore the contents of executables
+    // and execution regions (they don't access globals). i.e.
+    //   explorer.setOpAction<IREE::Stream::ExecutableOp>(TraversalAction::IGNORE);
+    //   explorer.setOpAction<IREE::HAL::ExecutableOp>(TraversalAction::IGNORE);
+    explorer.initialize();
+
+    assert(rootOp->getNumRegions() == 1 && "expected module-like root op");
+
+    // Populate the list of globals to optimize. This is primarily to force the
+    // order in which globals are processed to be deterministic, reducing minor
+    // ordering variations in the resulting IR.
+    rootOp->walk([&](IREE::Util::GlobalOpInterface global) {
+      globalNames.push_back(global.getGlobalName());
+    });
+  }
+
+  AsmState &getAsmState() { return solver.getAsmState(); }
+  Explorer &getExplorer() { return explorer; }
+
+  // Runs analysis and populates the state cache.
+  // May fail if analysis cannot be completed due to unsupported or unknown IR.
+  LogicalResult run() {
+    // Initialize all potential accessors.
+    explorer.walk([&](Operation *op) {
+      if (isa<IREE::Util::GlobalLoadOpInterface>(op) ||
+          isa<IREE::Util::GlobalStoreOpInterface>(op) ||
+          isa<IREE::Util::GlobalLoadIndirectOpInterface,
+              IREE::Util::GlobalStoreIndirectOpInterface>(op) ||
+          isa<mlir::CallOpInterface>(op) ||
+          isa<mlir::CallableOpInterface>(op) ||
+          op->hasTrait<OpTrait::IREE::Util::YieldPoint>() ||
+          op->hasTrait<OpTrait::IREE::Util::ObjectLike>() ||
+          op->getNumRegions() > 0) {
+        solver.getOrCreateElementFor<GlobalAccessor>(
+            Position::forOperation(op));
+      }
+      return WalkResult::advance();
+    });
+
+    // Run solver to completion.
+    auto result = solver.run();
+    LLVM_DEBUG(solver.print(llvm::dbgs()));
+
+    // Populate the accessor map and other cached state for use when
+    // propagating direct accessors.
+    if (succeeded(result)) {
+      explorer.walk([&](Operation *op) {
```

Comment:
explorer.forEachGlobal here too to avoid the walk

Reply:
I just iterated over the list of elements stored by the solver instead (as that's essentially what I wanted here).